src/lib/storage/table.cpp

```diff
@@ -1,89 +1,111 @@
 #include "table.hpp"
 
+#include "resolve_type.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
-Table::Table(const ChunkOffset target_chunk_size) {
-  // Implementation goes here
+Table::Table(const ChunkOffset target_chunk_size) : _chunk_size{target_chunk_size} {
+  create_new_chunk();
 }
 
 void Table::add_column_definition(const std::string& name, const std::string& type, const bool nullable) {
-  // Implementation goes here
+  _column_names.emplace_back(name);
+  _column_types.emplace_back(type);
+  _is_column_nullable.emplace_back(nullable);
+}
+
+static void add_value_segment(Chunk& chunk, const std::string& type, const bool nullable) {
+  resolve_data_type(type, [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    const auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(nullable);
+    chunk.add_segment(std::move(value_segment));
+  });
 }
 
 void Table::add_column(const std::string& name, const std::string& type, const bool nullable) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(row_count() == 0, "Tried to create column on non-empty table.");
+  add_column_definition(name, type, nullable);
+  add_value_segment(*_chunks.back(), type, nullable);
 }
 
 void Table::create_new_chunk() {
-  // Implementation goes here
+  _chunks.emplace_back(std::make_shared<Chunk>());
+  for (ColumnID id{0}; const auto& type : _column_types) {
+    add_value_segment(*_chunks.back(), type, _is_column_nullable[id]);
+    ++id;
+  }
 }
 
 void Table::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (_chunks.back()->size() == target_chunk_size()) {
+    create_new_chunk();
+  }
+  _chunks.back()->append(values);
 }
 
 ColumnCount Table::column_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _chunks.front()->column_count();
 }
 
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return (chunk_count() - 1) * target_chunk_size() + _chunks.back()->size();
 }
 
 ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return static_cast<ChunkID>(_chunks.size());
 }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  for (ColumnID id{0}; const auto& name : _column_names) {
+    if (name == column_name) {
+      return ColumnID{id};
+    }
+    ++id;
+  }
+  Fail("Tried to find a non-existent column.");
```

Comment:
I am not sure I understand what the assertion would look like. We want to make sure that a matching column is found which is equivalent to not reaching the code after the loop. Do you mean to replace this with `Assert(false, ...)`? Because this is then just the same as `Fail(...)`, but I perceive it as more noisy.